File: solutions/747.min-cost-climbing-stairs/min-cost-climbing-stairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        // int[] dp = new int[cost.length + 1];
        // for(int i = 2; i < dp.length; i++) 
        //     dp[i] = Math.min(dp[i - 1] + cost[i - 1], dp[i - 2] + cost[i - 2]);
        // return dp[cost.length];
        for(int i = 2; i < cost.length; i++) 
            cost[i] += Math.min(cost[i - 1], cost[i - 2]);
        return Math.min(cost[cost.length - 1], cost[cost.length - 2]);
    }
}

File: solutions/747.min-cost-climbing-stairs/min-cost-climbing-stairs.cpp
class Solution {
public:
    // int minCostClimbingStairs(vector<int>& cost) {
    //     int dp1 = 0, dp2 = 0, dp = 0; 
    //     for(int i = 1; i < cost.size(); i++) {
    //         dp = min(dp1 + cost[i], dp2 + cost[i - 1]);
    //         dp2 = dp1;
    //         dp1 = dp;
    //     }
    //     return dp;
    // }
    int minCostClimbingStairs(vector<int>& cost) {
        vector<int> m(cost.size() + 1, 0);
        return dp(m, cost, cost.size());
    }
private:
    int dp(vector<int>& m, vector<int>& cost, int i) {
        if(i < 2) return 0;
        if(m[i] > 0) return m[i];
        return m[i] = min(dp(m, cost, i - 1) + cost[i - 1], dp(m, cost, i - 2) + cost[i - 2]);
    }
};